feat: experimental sample-diagnosis endpoint

e-t-k opened the pull request:
Discussion: https://github.com/CBIIT/ccdi-federation-api/discussions/131 **Improved disease querying**
Projects:
https://github.com/orgs/CBIIT/projects/19/views/1?pane=issue&itemId=102925127
https://github.com/orgs/CBIIT/projects/19/views/1?pane=issue&itemId=102925260

**PR Close Date:** Not yet determined

**Main feature**
Adds `sample-diagnosis` endpoint with a `search` parameter. This parameter performs case-insensitive string subset searching on the `diagnosis` field. 
Examples:
[http://localhost:8000/sample-diagnosis?search=Random Diagnosis](http://localhost:8000/sample-diagnosis?search=Random%20Diagnosis) : Returns all samples with a diagnosis.
http://localhost:8000/sample-diagnosis?search=X : Returns all samples with diagnosis `Random Diagnosis X`
[http://localhost:8000/sample-diagnosis?search=SIS y](http://localhost:8000/sample-diagnosis?search=SIS%20y) Returns all samples with diagnosis `Random Diagnosis Y`

Other `sample` parameters are also usable with `sample-diagnosis`. 
Example: http://localhost:8000/sample-diagnosis?tissue_type=Tumor&per_page=2

**Other changes**
To support the main feature, this pull also implements the following:
- Adds missing `sample?diagnosis=xyz` search parameter. Previously, searching by diagnosis would return all samples
regardless of query value. Example: http://localhost:8000/sample?diagnosis=Random%20Diagnosis%20A
- Updates the sample's diagnosis randomizer. Previously every sample had the same disease, "Random Diagnosis";
now it is "Random Diagnosis X" where X is a letter or digit. This enables diagnosis searching to return meaningful results.
- Adds missing `sample/by/diagnosis/count` endpoint. Example: http://localhost:8000/sample/by/diagnosis/count

**Known Caveats**
- `sample-diagnosis` does not yet appear in the swagger.yml file
- the `search` parameter is not limited to the diagnosis character set described in #131 and will allow searches on all sorts of characters rather than returning an error. Example: [http://localhost:8000/sample-diagnosis?search=/Example🙂%^](http://localhost:8000/sample-diagnosis?search=/Example%F0%9F%99%82%^)
- This implementation also adds the `search` parameter to the `sample` endpoint itself, but testers can choose not to use it.
 
Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [X] You have added the relevant groups/individuals to the reviewers.
- [ ] Your commit messages conform to the [Conventional
      Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard. *(I missed one!)*
- [x] You have updated the README or other documentation to account for these
      changes (when appropriate). *(Not applicable for this change)*
- [x] You have added a line describing the change in the `CHANGELOG.md` under
      `[Unreleased]`.

grugna commented on crates/ccdi-server/src/filter/sample.rs:
```diff
@@ -36,82 +38,116 @@ impl FilterMetadataField<Sample, FilterSampleParams> for Vec<Sample> {
+
+                    // If no metadata is included, this entry should not be
+                    // included.
+                    false
```
I assume you meant `return false`?

e-t-k replied:
@grugna Yes, I would assume that too! But when I changed it to `return false;` just now I got an error  from `cargo clippy -- -D warnings` linking to the [Clippy Lints: needless return](https://rust-lang.github.io/rust-clippy/master/index.html#needless_return) page.

So, I will leave it as `false` so that this code is more "rusty". Thanks for checking.

grugna commented:
Since you mentioned you would have made the `subject-diagnosis endpoints` in a different PR could you just rename this one from `feat: experimental sample-diagnosis and subject-diagnosis endpoints` to `feat: experimental sample-diagnosis`?

mcrusch commented:
> @e-t-k I see a query parameter "search" was added to "sample" endpoint to improve disease querying by name patterns. Since this disease querying returns the same entities as other sample querying requests by parameters, do we opt for now to just having this extra "sample" parameter instead of creating a new endpoint "sample-diagnosis"?
> 
> When Michael suggested adding new endpoints one of the arguments was not to touch the existed implementation v1.1.0. Adding an extra parameter might be an easier way than adding a new endpoint so. If we include "search" parameter to "sample" would it be a commitment to implement this querying in "sample"? Thank you.

I would advocate for keeping it as a separate endpoint for now for a few reasons:

1. Currently this is a one-off querying method, inconsistent with how all of the other querying works.
2. I don't think this search method is the ideal end state.  If we were all harmonized to a single tree, then it would probably make more sense to do exact search based on code for the nodes in the tree and retrieve everything under that node.  Or, if we were going to stick with text-based, then we might want to consider extending that to other elements, or using a more sophisticated full text search with stemming to account for different spelling.  I'd prefer to hold off on integrating into the main endpoint until it's more finalized.
3. Changes to the design at this point would add risk to the timeline.

mcrusch commented:
I reviewed the example API, and it looked correct and worked well.  I don't see the new endpoint in the swagger docs, though.  Was that intentional because it's experimental?  I was thinking it would be in there, and that would be the place where we explain the fact that it's experimental and subject to change.

mcrusch commented:
I double-checked my review, and I missed some things before.

It looks to me like there is both a new endpoint `sample-diagnosis` that accepts the `search` query, AND also the `search` query is supported on the `sample` endpoint.  I think we only want `search` on the new endpoint, not the `sample` endpoint.

Also, the Swagger doc shows the new `search` query under `sample`, but doesn't show the new endpoint.

asafieva commented on swagger.yml:
"diagnosis" search query parameter is added all right.
"search" query parameter added is not expected in "sample" endpoint.
There is no "sample-diagnosis" endpoint added.

e-t-k replied:
Thank you! I have fixed these.

asafieva reviewed (changes requested):
Please remove "sample"."search" query parameter since it shall be in a separate endpoint.
The changes related to "sample"."diagnosis" query parameter are OK to merge.

mcrusch commented on crates/ccdi-server/src/params/filter.rs:
```diff
@@ -161,6 +163,18 @@ pub struct Sample {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    #[param(required = false, nullable = false)]
+    pub diagnosis: Option<String>,
+
+    /// Matches any sample where the `diagnosis` field contains the
+    /// string provided, ignoring case.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    #[param(required = false, nullable = false)]
+    pub search: Option<String>,
```
By adding "search" here, it is automatically available to the regular `sample` endpoint.  To restrict it to the experimental endpoint, it needs to be in a separate struct.  Clay recommended using composition + serde flatten.  Basically, a new struct after Sample called SampleDiagnosis or SampleExperimental that has a Sample, but is flattened in the serialized form.  So, perhaps something like (sorry, my rust is not good):
```suggestion
}

#[derive(Debug, Default, Deserialize, IntoParams, Introspect, Serialize)]
#[into_params(parameter_in = Query)]
pub struct SampleDiagnosis {
    #[serde(flatten)]
    pub sample: Sample,

    /// Matches any sample where the `diagnosis` field contains the
    /// string provided, ignoring case.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[param(required = false, nullable = false)]
    pub search: Option<String>,
```

e-t-k replied:
This is really helpful. I wasn't confident in my understanding of how to use composition so I ended up copy-pasting the existing Sample query parameters into SampleDiagnosis instead. But something to look into for Subject!

mcrusch commented on crates/ccdi-server/src/routes/sample.rs:
To keep the new search functionality separate, recommend creating a new file for the route sample_diagnosis.rs (or sample-diagnosis.rs?).  The new route would use the SampleDiagnosis filter instead of the Sample filter to get the extra parameter.  Clay said the macros are hard to inherit so the easiest way to do this would actually be to copy sample.rs to sample_diagnosis.rs and then update it.

So, the only update to sample.rs would be to add the exact querying for diagnosis.  That plus the other changes would be in sample_diagnosis.rs.

Finally, to get the new endpoint to show up in the swagger doc, add it to crates/ccdi-openapi/src/api.rs.  I think we should create a new section there for experimental endpoints and group the two new endpoints that way.  Alternatively, they could go under the sample and subject sections, respectively.

e-t-k replied:
Thanks so much for this guidance! It was incredibly helpful, especially the pointer to `api.rs` to update the swagger doc.

e-t-k commented:
Hi all, thanks so much for the discussion. I've addressed the changes requested. At a high level:
 - The `search` query parameter has been removed from `sample` endpoint.
 - The new endpoint `sample-diagnosis` has been added to the swagger docs under an **Experimental** header.
 - Per [@grugna's suggestion](https://github.com/CBIIT/ccdi-federation-api/pull/139#issuecomment-2847995501), updated title to mention sample-diagnosis only.

Since this most recent commit changed a lot of code, I've added a quick overview of the changes to each file in [the commit text](https://github.com/CBIIT/ccdi-federation-api/pull/139/commits/ff33f02e5edc6a70347d0e882f21de64b0a96531) -- please take a look.

mcrusch reviewed (approved):
Tested again with the latest changes.  Swagger doc and API testing both looked good: search present and working in sample-diagnosis, but absent from sample.  I approve.

asafieva commented on swagger.yml:
@e-t-k swagger.yml "Experimental" tag shall be added to "tags" YAML section (Ln 19017).

e-t-k replied:
@asafieva resolved in https://github.com/CBIIT/ccdi-federation-api/pull/139/commits/d92ac10bd98b0ea783c5561118b0b0cfbed7726e , thanks for catching!
